### PR TITLE
v0.0.2, add cookie mode

### PR DESCRIPTION
**Pull request by kingiw**

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @GopherJ please review.

**Comment by hsluoyz** [comment]
@kingiw please fix DCO.

File: example/react/Readme.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Casbin.js in React
```
Comment (hsluoyz) [inline example/react/Readme.md:1]: This library should be kept as small as possible, so no example code here. The examples can be put into: https://github.com/casbin-js

File: package.json [hunk 2]
```diff
@@ -37,6 +38,8 @@
     "webpack-merge": "^5.0.8"
   },
   "dependencies": {
-    "axios": "^0.19.2"
+    "@types/js-cookie": "^2.2.6",
```
Comment (nodece) [inline package.json:41]: `@types/js-cookie` should be as devDependencies.

File: src/Authorizer.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    /**
+     * 
+     * @param mode "auto", "cookies" or "manual"
+     * "auto": Specify the casbin server endpoint, and Casbin.js will load permission from it when the identity changes
+     * "cookies": Casbin.js load the permission data from the cookie "casbin_permission" or the specified cookie key.
+     * "manual": Load the permission mannually with "setPermission"
+     * @param args.endpoint Casbin service endpoint, required when mode == "auto"
+     * @param args.cookieKey The cookie key when loading permission, activated when mode == "cookies"
+     */
+    constructor(mode: Mode, args: {endpoint?: string, cookieKey?: string} = {}) {
```
Comment (nodece) [inline src/Authorizer.ts:29]: Here are set default value for mode is better: `constructor(mode: Mode = 'manual', args: {endpoint?: string, cookieKey?: string} = {}) `

File: src/Authorizer.ts [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        if (mode == 'auto') {
+            if (!args.endpoint) {
+                throw new Error("Specify the endpoint when initializing casbin.js with mode == 'auto'");
+                return;
```
Comment (nodece) [inline src/Authorizer.ts:33]: `return` is unecessary.

File: src/Authorizer.ts [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+    public setPermission(permission : Record<string, unknown> | string) : void{
+        this.permission.load(permission);
+    }
+
+    /**
+     * Get the authority of a given user from Casbin core
+     */
+    private async syncUserPermission(): Promise<void> {
+        if (this.endpoint !== undefined) {
```
Comment (nodece) [inline src/Authorizer.ts:68]: we also consider `this.endpoint` is `null`.

File: example/react/src/CasbinDemo.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({user: nextProps.user});
+  }
+
+  render() {
+    // set the permission when the user identity changes.
+    this.auth.setPermission(permissions[this.state.user]);
+    const alice_div = <p>This is alice's data.</p>
```
Comment (nodece) [inline example/react/src/CasbinDemo.js:21]: please use Camel-Case: `const AliceDiv = <p>This is alice's data.</p>`

File: example/react/src/CasbinDemo.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    const bob_div = <p>This is bob's data.</p>
```
Comment (nodece) [inline example/react/src/CasbinDemo.js:22]: +1

File: example/react/src/CasbinDemo.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    return (
+      <div>
+        <p>====</p>
+        { this.auth.can('read', 'alice_data') && alice_div }
```
Comment (nodece) [inline example/react/src/CasbinDemo.js:26]: `alice_div` is component, should be use `<AliceDiv>`.

File: example/react/src/CasbinDemo.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        { this.auth.can('read', 'bob_data') && bob_div}
```
Comment (nodece) [inline example/react/src/CasbinDemo.js:27]: +1

File: example/react/src/CasbinDemo.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+    this.auth = new Authorizer("manual");
+  }
+
+  componentWillReceiveProps(nextProps) {
+    this.setState({user: nextProps.user});
+  }
+
+  render() {
+    // set the permission when the user identity changes.
+    this.auth.setPermission(permissions[this.state.user]);
```
Comment (nodece) [inline example/react/src/CasbinDemo.js:20]: consider network request, please move the code to `componentDidMount`.

File: example/react/src/CasbinDemo.js (hunk 8 above)
Comment (kingiw) [reply example/react/src/CasbinDemo.js:26]: Nope, AliceDiv is a JSX Element and it should used like this.

File: example/react/src/CasbinDemo.js (hunk 10 above)
Comment (kingiw) [reply example/react/src/CasbinDemo.js:20]: Nope, this is improper. `this.auth.setPermission()` here will not trigger any network request.

Maybe the react example doesn't properly designed and cannot clearly show the usage (and the mechanism) here. I will update this example later.

File: example/react/Readme.md (hunk 1 above)
Comment (kingiw) [reply example/react/Readme.md:1]: I will separate this example into an independent repo.

**Comment by kingiw** [comment]
I've removed the example files from the PR

File: example/react/src/CasbinDemo.js (hunk 6 above)
Comment (hsluoyz) [reply example/react/src/CasbinDemo.js]: I think `aliceDiv` is better?

File: example/react/src/CasbinDemo.js (hunk 10 above)
Comment (hsluoyz) [reply example/react/src/CasbinDemo.js]: render() is for display flow. We shouldn't "set" something in render(), right?

**Comment by hsluoyz** [comment]
> I've removed the example files from the PR

Please replace existing commits. The example should not show up in commits.

**Comment by nodece** [comment]
@kingiw  You should update the PR title.

**Pull request merged**